### PR TITLE
Users action

### PR DESCRIPTION
**Pull request by IdanTo1**
a. implement all user related Actions
b. add all requested methods to Session and User

### CODE REVIEW

File: src/Action.cpp [hunk 1]
```diff
@@ -23,60 +19,124 @@ const std::string BaseAction::getStatusString() const
...
+    }
+    else {
+        error(INVALID_ALG_ERR);
+        return;
+    }
+    complete();
+}
+std::string CreateUser::toString() const {
+    return ("creating new User - " + _userName); //TODO complete due to forum answer
+}
```
Comment (Hroy02) [inline src/Action.cpp]: toString() isn't what the Action prints. it's the description of the Action. It's what should come up in the printActionsLog.
So here it should be `"CreateUser" + getStatusString()`

File: include/Action.h [hunk 2]
```diff
@@ -9,57 +9,72 @@ class Session;
 enum ActionStatus{
 	PENDING, COMPLETED, ERROR
 };
-
+const std::string USER_EXISTS_ERR = "User already exists";
+const std::string INVALID_ALG_ERR = "Invalid algorithm was given";
+const std::string USER_DOESNT_EXISTS_ERR = "Requested user does not exists";
```
Comment (Hroy02) [inline include/Action.h]: does not exist*

File: src/Action.cpp [hunk 3]
```diff
@@ -23,60 +19,124 @@ const std::string BaseAction::getStatusString() const
...
+ChangeActiveUser::ChangeActiveUser(const std::string& name): BaseAction(), _newUser(name) {}
 
+void ChangeActiveUser::act(Session &sess) {
+    const std::unordered_map<std::string,User*>& usersMap = sess.getUsers();
+    if(!usersMap.count(_newUser)) {
+        error(USER_DOESNT_EXISTS_ERR);
+    }
+    else {
+        sess.setActiveUser(*(usersMap.at(_newUser)));
+        complete();
```
Comment (Hroy02) [inline src/Action.cpp:72]: using find instead of count here would mean you could iterate the map only once

File: src/Action.cpp [hunk 4]
```diff
@@ -23,60 +19,124 @@ const std::string BaseAction::getStatusString() const
...
     }
 }
 
-void Exit::act(Session& sess)
-{
+std::string ChangeActiveUser::toString() const {
+    return "Active user changed to" + _newUser; //TODO
+}
```
Comment (Hroy02) [inline src/Action.cpp]: Again, should be `"ChangeActiveUser" + getStatusString()`

File: src/Action.cpp [hunk 5]
```diff
@@ -23,60 +19,124 @@ const std::string BaseAction::getStatusString() const
...
+    }
+    else {
+        sess.deleteUserFromMap(_userName);
+        complete();
+    }
+}
+
+std::string DeleteUser::toString() const {
+    return ("Deleting user" + _userName); // TODO
+}
```
Comment (Hroy02) [inline src/Action.cpp]: Again

File: src/Action.cpp [hunk 6]
```diff
@@ -23,60 +19,124 @@ const std::string BaseAction::getStatusString() const
...
+    else {
+        User& oldUser = *(usersMap.at(_oldUserName));
+        User* newUser = oldUser.clone();
+        sess.addToUserMap(newUser);
+    }
+}
+
+std::string DuplicateUser::toString() const {
+    return ("Created new user" + _newUserName + "from existing user" + _oldUserName); // TODO
+}
```
Comment (Hroy02) [inline src/Action.cpp]: Again

File: src/Action.cpp [hunk 7]
```diff
@@ -23,60 +19,124 @@ const std::string BaseAction::getStatusString() const
...
+    if(!usersMap.count(_oldUserName)) {
+        error(USER_DOESNT_EXISTS_ERR);
+    }
+    else if (usersMap.count(_newUserName)) {
+        error(USER_EXISTS_ERR);
+    }
+    else {
+        User& oldUser = *(usersMap.at(_oldUserName));
+        User* newUser = oldUser.clone();
+        sess.addToUserMap(newUser);
```
Comment (Hroy02) [inline src/Action.cpp:112]: Here I also think that map::find(_oldUserName) would be more efficient

File: src/Session.cpp [hunk 8]
```diff
@@ -21,6 +21,7 @@ std::vector<std::string> Session::extractTags(nlohmann::json& tagList)
 
 void Session::fillContentFromJson(const std::string &configFilePath)
 {
+    content = std::vector<Watchable*>();
```
Comment (Hroy02) [inline src/Session.cpp]: YOU commented that this can be removed because it is already initialized anyway :)

**Review by Hroy02 — CHANGES_REQUESTED** [review]
Most are minor
The big one is the Action::toString()

File: src/Action.cpp (hunk 1 above)
Comment (IdanTo1) [reply src/Action.cpp]: and I was trying so hard :cry:

File: src/Action.cpp (hunk 3 above)
Comment (IdanTo1) [reply src/Action.cpp:72]: don't understand. count is
Average case: linear in the number of elements counted.
Worst case: linear in container size.
 - from cpp reference.
so its between O(1)-O(n). sounds good to me...

File: src/Action.cpp (hunk 7 above)
Comment (IdanTo1) [reply src/Action.cpp:112]: same

Comment (IdanTo1) [reply src/Action.cpp:112]: the name is confusing. it doesn't really count... it returns 0 or 1. its basically "contain". find returns an iterator and so less efficient

File: src/Session.cpp (hunk 8 above)
Comment (IdanTo1) [reply src/Session.cpp]: auto merge :see_no_evil: 
nice you noticed

File: src/Action.cpp (hunk 3 above)
Comment (Hroy02) [reply src/Action.cpp:72]: Complexity won't change. But finding the same element twice (once with count and once with at) seems quicker, albeit not **much** quicker

File: src/Action.cpp [hunk 9]
```diff
@@ -74,7 +74,7 @@ void ChangeActiveUser::act(Session &sess) {
 }
 
 std::string ChangeActiveUser::toString() const {
-    return "Active user changed to" + _newUser; //TODO
+    return ("changeuser" + getStatusString()); //TODO
```
Comment (Hroy02) [inline src/Action.cpp]: Maybe `ChangeUser` just to fit with the example in the pdf?

**Review by Hroy02 — APPROVED** [review]
Perfect

**Pull request merged**